boltdesignsystem/bolt
# Epic - Typeahead: CSS Modules

## Jira
http://vjira2:8080/browse/BDS-1807

## Summary
Adds CSS Module support to the Build Tools. These updates allow for auto-generated scoped CSS selectors to keep templates, styles and JS all in sync. 

Part of the broader [Typeahead](#1311) component work.

## Details
- Adds opt-in support for [CSS Modules](https://css-tricks.com/css-modules-part-1-need/) to the existing Build process. 
- Only components with Sass files named `NAME.scoped.scss` have their class names semi-randomly generated when compiling. 
- The way these build updates are set up, the Sass file ending in `.scoped.scss` is what triggers the JSON file with the key:value pairs to get generated and added to the folder the Sass file is living in. This is the critical piece in how the class names that get used in the Twig template stay in sync with the CSS getting compiled.
- Javascript files pulling in `.scoped.scss` files also trigger this "opt-in" CSS Module behavior, however these use a different mechanism under the hood to generate the generated class names. Essentially we're using two different off the shelf approaches to handling CSS Modules that both share the same "randomly generated class name" selector logic.

## How to test
- Review PR changes
- Confirm Travis build passing as expected
- Check out the full Typeahead code branch to see how these changes actually fit into the current workflow + build process:
  - [typeahead.scoped.json](https://github.com/bolt-design-system/bolt/blob/feature/typeahead/packages/components/bolt-typeahead/typeahead.scoped.json)
  - [typeahead.scoped.scss](https://github.com/bolt-design-system/bolt/blob/feature/typeahead/packages/components/bolt-typeahead/typeahead.scoped.scss) (nothing different other than the `.scoped.scss` file name)
  - [typeahead.twig](https://github.com/bolt-design-system/bolt/blob/feature/typeahead/packages/components/bolt-typeahead/typeahead.twig#L1) -- pulls in the external (dynamically generated but still committed to Git) JSON file to map to the compiled CSS class names / selectors
  - [typeahead.js](https://github.com/bolt-design-system/bolt/blob/feature/typeahead/packages/components/bolt-typeahead/typeahead.js#L10) - which pulls in the scoped Sass file + maps to a modified version of the classnames helper library to match to selectors [like this](https://github.com/bolt-design-system/bolt/blob/feature/typeahead/packages/components/bolt-typeahead/typeahead.js#L104) automatically
